core/lib/config/src/configs/chain.rs

```diff
@@ -45,6 +45,18 @@ impl Default for FeeModelVersion {
     }
 }
 
+#[derive(Debug, Clone, Copy, Deserialize, PartialEq, Eq)]
+pub enum L1BatchCommitDataGeneratorMode {
+    Rollup,
+    Validium,
+}
+
+impl Default for L1BatchCommitDataGeneratorMode {
+    fn default() -> Self {
+        Self::Rollup
+    }
+}
```

Comment:
I've taken a look at the PR to the contracts and it seems that info [on whether we're in validium mode is available on the contract](https://github.com/lambdaclass/era-contracts/pull/14/files#diff-6466768dc18a351a71e348fcf8c13a415a9aeb9d676b6d671ebd61660ab90f53R27).

Can we increase the safety by querying the contract and panicking if the chosen config option doesn't match what's enabled on the contract?

This probably would require you to create a corresponding getter in contracts, but I guess it isn't too hard.

Reply:
It will be less complex to read the `VALIDIUM_MODE` env as it is the same that the contract reads (no more contract logic is needed nor making a query to it). Any thoughts on this? This will also reduce the code complexity as the config and enum won't be required any longer.

Reply:
The contracts are deployed and the server is run usually by different people at different times (and moreover, configs for server can change over time), so it's not guaranteed that this env variable would have the same value all the time, so this would serve as a defensive programming measure to guarantee that we won't ever run server in incompatible mode by accident.

Reply:
hmm @ilitteri so what you mean is to read env directly from one of the `L1BatchCommitDataGeneratorMode` structs and make sure that e.g. if Rollup one is instantiated, the env var is set to `Rollup`? But we define which `L1BatchCommitDataGeneratorMode` to create based on this very config variable! I mean, we could mess up env parsing into variable, but it's not worth it to check again in run-time :)

What @popzxc suggests is to panic if we mistakenly run server configured in rollup mode against contracts that are validium.

Reply:
> hmm @ilitteri so what you mean is to read env directly from one of the L1BatchCommitDataGeneratorMode structs and make sure that e.g. if Rollup one is instantiated, the env var is set to Rollup? But we define which L1BatchCommitDataGeneratorMode to create based on this very config variable! I mean, we could mess up env parsing into a variable, but it's not worth it to check again in run-time :)

I'm saying that rolling back to [this](https://github.com/lambdaclass/zksync-era/pull/75#discussion_r1474245028) will be easier in terms of complexity (if this solution fits), and given that there's only one `VALIDIUM_MODE` env variable, we are sure that its value is the same for the contracts and the server. This way we also do not need to panic because this env variable is set in the `init.ts`. 

Let me know if I'm missing something and if there's still a chance of having mixed configurations.

Reply:
I think our point is that `VALIDIUM_MODE` may have different values at different times (e.g. you could deploy contracts with `VALIDIUM_MODE=true` and then run the server with `VALIDIUM_MODE=false` by accident, which would be a mistake during the configuration.

So the proposal here is not about simplification, but rather about making the code more resistant to incorrect configuration. You will still have an option for the mode (and it doesn't matter much how is it called) -- it will express the desire of the node operator to run the node with a certain config, but at the same time the node itself will check whether it was properly configured and will not let the node to run in an incorrect mode.

Reply:
Just in case, @ilitteri if you do decide to create such a getter, please make sure that it is marked as unstable in the comments. In the future we might wanna allow different chains to have different DA layers (i.e. not a binary "Rollup"/"Validium") and so a different method so a different double checking method will be required.

Though I am all for adding a temporary getter for double checking to avoid mistakes

Reply:
Coming back to this conversation, we've recently merged this PR https://github.com/lambdaclass/zksync-era/pull/138 which makes the `zk init` command update the config files automatically correspondingly to whether the `--validium-mode` flag is set or not, leaving the possibility of having mix configurations to the only case of deliberately changing the `dev.env` file before running `zk server`. The `VALIDIUM_MODE` env variable is only used for solpp in the L1 contracts compilation. There is no way to incorrectly set up the whole stack using the `zk` tool.

Reply:
@ilitteri the `zk` tool is only used to initialize the network for the first run. However, once the system in production, configs live elsewhere (not in `dev.env`) and are often edited. For example, we have dedicated infrastructure for managing configuration of our environments. So the problem we're discussing may occur *long after* the chain is initialized when it's in production. The check, if added, would happen each time the node is restarted (and you have to restart the node when you change configuration), so it'll act as a circuit breaker not allowing it to start if somehow somebody changed the configuration to an incorrect setting.